src/java.base/share/classes/java/time/ZoneOffset.java

```diff
@@ -507,6 +507,11 @@ public ZoneRules getRules() {
         return ZoneRules.of(this);
     }
 
+    @Override
+    /* package-private */ ZoneOffset getOffset(long epochSecond, int nanoOfSecond) {
+        return this;
```

Comment:
An alternate approach would be for `ZoneOffset` to cache the instance of `ZoneRules` either on construction or first use (racy idiom would be OK). That way this issue is solved for the many different places that call `zoneId.getRules()`.

Reply:
Why not both? I measured an improvement using that alone, but specifically avoiding going via getRules is faster still (without adversely affecting `ZoneRegion` paths). I've added the cache in `ZoneOffset`, along with an override of `ZoneId::normalized` in `ZoneOffset` to shortcut the `getRules`.